overlay.go

```diff
@@ -331,3 +338,331 @@ func isZero(v reflect.Value) bool {
+	if series := overlay.Data.Series; series != "" {
+		base.Data.Series = series
+	}
+
+	// Append any additional relations.
+	base.Data.Relations = append(base.Data.Relations, overlay.Data.Relations...)
+
+	// Override machine definitions.
+	if machines := overlay.Data.Machines; machines != nil {
+		base.Data.Machines = machines
+	}
+
+	return nil
+}
+
+// removeRelations removes any relation defined in data that references
+// the application appName.
+func removeRelations(data [][]string, appName string) [][]string {
+	var result [][]string
+	for _, relation := range data {
+		// Keep the dud relation in the set, it will be caught by the bundle
+		// verify code.
+		if len(relation) == 2 {
+			left, right := relation[0], relation[1]
+			if left == appName || strings.HasPrefix(left, appName+":") ||
+				right == appName || strings.HasPrefix(right, appName+":") {
+				continue
+			}
+		}
+		result = append(result, relation)
+	}
+	return result
+}
+
+// mergeStructs iterates the fields of srcStruct and merges them into the
+// equivalent fields of dstStruct using the following rules:
+//
+// - if src defines a value for a scalar or slice field, it will overwrite
+//   the value from the dst (e.g. trust, series etc).
+// - if the src specifies a nil/empty value for a map field, then the map
+//   field of dst will be cleared.
+// - if the src specifies a non-empty value for a map field, its key/value
+//   tuples are iterated and:
+//   - if the value is nil/zero, it is deleted from the dst map.
+//   - otherwise, the key/value is inserted into the dst map overwriting any
+//     existing entries.
+func mergeStructs(dstStruct, srcStruct interface{}, fpm FieldPresenceMap) {
+	dst := reflect.ValueOf(dstStruct)
+	src := reflect.ValueOf(srcStruct)
+	typ := src.Type()
+
+	// Dereference pointers
+	if src.Kind() == reflect.Ptr {
+		src = src.Elem()
+		typ = src.Type()
+	}
+	if dst.Kind() == reflect.Ptr {
+		dst = dst.Elem()
+	}
+	dstTyp := dst.Type()
+
+	// Sanity check
+	if typ.Kind() != reflect.Struct || typ != dstTyp {
+		panic(errors.Errorf("BUG: source/destination type mismatch; expected destination to be a %q; got %q", typ.Name(), dstTyp.Name()))
+	}
+
+	for i := 0; i < typ.NumField(); i++ {
+		// Skip non-exportable fields
+		structField := typ.Field(i)
+		srcVal := src.Field(i)
+		if !srcVal.CanInterface() {
+			continue
+		}
+
+		fieldName := yamlName(structField)
+		if !fpm.fieldPresent(fieldName) {
+			continue
+		}
+
+		switch srcVal.Kind() {
+		case reflect.Map:
+			// If a nil/empty map is provided then clear the destination map.
+			if isZero(srcVal) {
+				dst.Field(i).Set(reflect.MakeMap(srcVal.Type()))
+				continue
+			}
+
+			dstMap := dst.Field(i)
+			if dstMap.IsNil() {
+				dstMap.Set(reflect.MakeMap(srcVal.Type()))
+			}
+			for _, srcKey := range srcVal.MapKeys() {
+				// If the key points to an empty value delete it from the dst map
+				srcMapVal := srcVal.MapIndex(srcKey)
+				if isZero(srcMapVal) {
```

Comment:
Does this mean that:
{
 "foo": 0
}

will delete foo?
I'm wondering about config items where you might want to actually set a value to 0. Thoughts?
I'm just wondering about cases where there is a default value, so setting the value to 0 or "" is different than not setting it at all.

Reply:
That is correct and I hadn't really thought about this case. However we _do_ need this functionality for _some_ of the maps (`offers` within `ApplicationSpec` being one example). 

I wonder if we could address this problem by checking whether the value is a scalar and only applying the delete logic for object values. What do you think?